### PR TITLE
chore: Session replay tests [wip]

### PR DESCRIPTION
**Pull request by metal-messiah**
This is a [wip] PR for adding jest and wdio tests around the (future, experimental) SR feature.  Do not review until marked as ready
---

### Overview
[WIP]

### Related Issue(s)
NEWRELIC-8807
NEWRELIC-8808

### Testing
Jest "component" test has been added
Several WDIO tests added [wip]

### CODE REVIEW

**Comment by metal-messiah** [comment]
JIL test suite --> https://github.com/newrelic/newrelic-browser-agent/actions/runs/5221260389/jobs/9425307817

**Comment by metal-messiah** [comment]
wdio suite --> https://github.com/newrelic/newrelic-browser-agent/actions/runs/5223315684/jobs/9430015981

File: src/features/session_replay/aggregate/index.js [hunk 1]
```diff
@@ -22,38 +22,41 @@ import { AggregateBase } from '../../utils/aggregate-base'
...
+
 let recorder, gzipper, u8
 
 /** The "mode" with which the session replay is recording */
-const MODE = {
+export const MODE = {
   OFF: 0,
   FULL: 1,
   ERROR: 2
 }
```
Comment (patrickhousley) [inline src/features/session_replay/aggregate/index.js:34]: Can we move constants to a `constants.js` in the root of the feature?

File: tests/specs/session-replay/harvest.e2e.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  describe('Session Replay Harvest Behavior', () => {
+    afterEach(async () => {
+      await browser.destroyAgentSession(browser.testHandle)
+    })
+
+    it('Should harvest early if exceeds preferred size', async () => {
+      const startTime = Date.now()
+      const [{ request: blobHarvest }] = await browser.url(await browser.testHandle.assetURL('rrweb-instrumented.html', config({ harvestTimeSeconds: 60 })))
+        .then(() => browser.waitForAgentLoad())
+        .then(() => Promise.all([
```
Comment (patrickhousley) [inline tests/specs/session-replay/harvest.e2e.js:13]: Can we avoid promise chains since we can use `async..await`?

File: tests/specs/session-replay/ingest.e2e.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      expect(await getSR()).toEqual(expect.objectContaining({
+        events: expect.any(Array),
+        initialized: true,
+        recording: true,
+        mode: 1,
+        blocked: false
+      }))
+
+      await Promise.all([
+        browser.testHandle.expectBlob(),
```
Comment (patrickhousley) [inline tests/specs/session-replay/ingest.e2e.js:34]: Is there a chance the initial SR harvest could happen before we get here?

File: tests/specs/session-replay/initialization.e2e.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+export default (function () {
+  describe('Session Replay Initialization', () => {
+    afterEach(async () => {
+      await browser.destroyAgentSession(browser.testHandle)
+    })
+
+    describe('Feature flags', () => {
+      // this test needs to be able to override the SR flag to 0
+      // work needed to enable this in WDIO
```
Comment (patrickhousley) [inline tests/specs/session-replay/initialization.e2e.js:12]: Is this comment still valid? Can you elaborate on what we are missing in WDIO to make this possible?

File: tests/specs/session-replay/payload.e2e.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    })
+
+    it('should be gzipped', async () => {
+      await browser.url(await browser.testHandle.assetURL('rrweb-instrumented.html', config()))
+        .then(() => browser.waitForAgentLoad())
+
+      const { request: harvestContents } = await browser.testHandle.expectBlob()
+
+      expect(harvestContents.query.content_encoding).toEqual('gzip')
+      expect(harvestContents.headers['content-encoding']).toEqual('gzip')
```
Comment (patrickhousley) [inline tests/specs/session-replay/payload.e2e.js:16]: Didn't we discover that setting this header caused the browser to do pre-flight CORS calls?

File: tests/specs/third-party-compatibility/jspdf.e2e.js [hunk 6]
```diff
@@ -1,22 +1,15 @@
...
-  withBrowsersMatching(supported)('2.5.1', async () => {
+  withBrowsersMatching(reliableUnload)('2.5.1', async () => {
     await runTest({
       browser,
       testAsset: 'third-party-compatibility/jspdf/2.5.1.html',
       afterLoadCallback: async () => {
         const [errorsResults] = await Promise.all([
           browser.testHandle.expectErrors(10000, true),
-          $('body').click() // Setup expects before interacting with page
+          browser.execute(function () { document.querySelector('body').click() }) // Setup expects before interacting with page
```
Comment (patrickhousley) [inline tests/specs/third-party-compatibility/jspdf.e2e.js:12]: What was the reason to switch this from the WDIO built-in to executing a function in the DOM?

File: src/features/utils/instrument-base.js [hunk 7]
```diff
@@ -85,19 +85,23 @@ export class InstrumentBase extends FeatureBase {
...
+  shouldImportAgg (featureName, session) {
   // if this isnt the FIRST load of a session AND
   // we are not actively recording SR... DO NOT run the aggregator
   // session replay samples can only be decided on the first load of a session
   // session replays can continue if in progress
-  if (featureName === FEATURE_NAMES.sessionReplay) return !!session?.isNew || !!session?.state.sessionReplayActive
-  // todo -- add case like above for session trace
-  return true
+    if (featureName === FEATURE_NAMES.sessionReplay) {
+      if (getConfigurationValue(this.agentIdentifier, 'session_trace.enabled') === false) return false
```
Comment (cwli24) [inline src/features/utils/instrument-base.js:101]: I think this a bad idea. Trace depends on replay already; don't introduce a cyclic relationship. They're becoming too coupled.

File: src/features/session_replay/aggregate/index.js (hunk 1 above)
Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:34]: I think Chun actually moved this in his PR to the session manager.  Will reflect once the dust settles

File: tests/specs/session-replay/helpers.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  return {
+    loader: 'experimental',
+    init: {
+      privacy: { cookies_enabled: true },
+      session_replay: { enabled: true, harvestTimeSeconds: 5, sampleRate: 1, errorSampleRate: 0, ...props }
+    }
+  }
+}
+
+export async function getSR () {
```
[the remark this answers is not in the document]
Comment (metal-messiah) [reply tests/specs/session-replay/helpers.js:20]: If we can figure out a way to make this "generic" -- ie get a feature by input param then yes we should absolutely expose it.  (or just make a handler for all the features).  This was super handy

File: tests/specs/session-replay/ingest.e2e.js (hunk 3 above)
Comment (metal-messiah) [reply tests/specs/session-replay/ingest.e2e.js:34]: not on our test server.  Only way for that to happen is if the initial payload exceeds 64kb compressed and our pages are way too small to have that happen

File: tests/specs/session-replay/initialization.e2e.js (hunk 4 above)
Comment (metal-messiah) [reply tests/specs/session-replay/initialization.e2e.js:12]: this comment needs to be removed, thanks

File: tests/specs/session-replay/payload.e2e.js (hunk 5 above)
Comment (metal-messiah) [reply tests/specs/session-replay/payload.e2e.js:16]: this header is only set by our testing server in the interceptor.  Should remove

File: src/features/utils/instrument-base.js (hunk 7 above)
Comment (metal-messiah) [reply src/features/utils/instrument-base.js:101]: Session replay literally cannot run if session trace is not included.  This is not optional

File: tests/specs/third-party-compatibility/jspdf.e2e.js (hunk 6 above)
Comment (metal-messiah) [reply tests/specs/third-party-compatibility/jspdf.e2e.js:12]: Deprecation warning in IDE.

File: .github/workflows/wdio-single-browser.yml [hunk 9]
```diff
@@ -81,7 +81,7 @@ jobs:
             -b ${{ inputs.browser-target }} \
             --concurrent 10 \
             ${{ runner.debug && '-v -L -D -d' }} \
-            ${{ inputs.coverage && '--coverage' }} \
+            ${{ inputs.coverage && '--coverage' || '' }} \
```
[the remark this answers is not in the document]
Comment (metal-messiah) [reply .github/workflows/wdio-single-browser.yml:84]: probably

File: tests/specs/third-party-compatibility/jspdf.e2e.js (hunk 6 above)
Comment (patrickhousley) [reply tests/specs/third-party-compatibility/jspdf.e2e.js:12]: The IDE is treating this as jQuery which it is not. The deprecation warning is wrong.

File: tests/specs/session-replay/ingest.e2e.js (hunk 3 above)
Comment (patrickhousley) [reply tests/specs/session-replay/ingest.e2e.js:34]: Should we try to make a test for pages that exceed the initial 64kb?

File: tests/specs/session-replay/session-state.e2e.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        // wait longer than harvest interval
+          await browser.testHandle.expectBlob(6000)
+          // expect bam endpoint to have only been called once (instead of twice)
+          fail('Should not have seen another blob request')
+        } catch (err) {
+          // if expectblob failed, this test passed.
+        } finally {
+          await browser.closeWindow()
+          await browser.switchToWindow((await browser.getWindowHandles())[0])
+        }
```
Comment (bjfield) [inline tests/specs/session-replay/session-state.e2e.js:127]:
```suggestion
        // Waiting for the second blob should time out, indicating no second call to the BAM endpoint.
        // The wait must be longer than harvest interval.
        await browser.testHandle.expectBlob(6000, true)
        await browser.closeWindow()
        await browser.switchToWindow((await browser.getWindowHandles())[0])
```

File: tests/specs/session-replay/session-state.e2e.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+export default (function () {
+  describe('session manager state behavior', () => {
+    afterEach(async () => {
+      await browser.destroyAgentSession(browser.testHandle)
+    })
+
+    describe('session manager mode matches session replay instance mode', () => {
+      it('should match in full mode', async () => {
```
Comment (bjfield) [inline tests/specs/session-replay/session-state.e2e.js:18]: This test passes only about half the time for me on my local, just FYI in case it matters.

File: tests/specs/session-replay/ingest.e2e.js (hunk 3 above)
Comment (metal-messiah) [reply tests/specs/session-replay/ingest.e2e.js:34]: good call.  we are only mocking this currently.  We should load a really big page for a better integration check

File: tests/specs/session-replay/session-state.e2e.js (hunk 10 above)
Comment (bjfield) [reply tests/specs/session-replay/session-state.e2e.js:127]: The purpose of this one is to utilize the expectTimeout flag instead of an expect error construct. It was logging the timeout error even though it was caught.

File: src/features/session_replay/aggregate/index.js [hunk 12]
```diff
@@ -68,58 +71,65 @@ export class Aggregate extends AggregateBase {
...
-        if (recorder && this.initialized) {
-          this.stopRecording()
-          this.startRecording()
-          this.scheduler.startTimer(this.harvestTimeSeconds)
-
-          const { session } = getRuntime(this.agentIdentifier)
-          session.state.sessionReplay = this.mode
+    const shouldSetup = (
+      getConfigurationValue(agentIdentifier, 'privacy.cookies_enabled') === true &&
+      getConfigurationValue(agentIdentifier, 'session_trace.enabled') === true
```
Comment (cwli24) [inline src/features/session_replay/aggregate/index.js:76]: Shouldn't cookies_enabled have already been checked for?

What's the intention with session_trace checking here? The config could be enabled, but it doesn't mean trace is actually running.

File: src/features/session_replay/aggregate/index.js [hunk 13]
```diff
@@ -68,58 +71,65 @@ export class Aggregate extends AggregateBase {
...
-    this.waitForFlags(['sr']).then(([{ value }]) => {
-      this.initializeRecording(
-        value,
-        Math.random() < getConfigurationValue(this.agentIdentifier, 'session_replay.errorSampleRate'),
-        Math.random() < getConfigurationValue(this.agentIdentifier, 'session_replay.sampleRate')
-      )
-    })
+      }, this.featureName, this.ee)
+
+      // new handler for waiting for multiple flags.  will be useful if/when backend designs multiple flags, or for evaluating multiple feature flags simultaneously (stn vs sr)
```
Comment (cwli24) [inline src/features/session_replay/aggregate/index.js:124]: Comment should prob go in `waitForFlags`

File: src/features/session_replay/aggregate/index.js [hunk 14]
```diff
@@ -152,6 +162,11 @@ export class Aggregate extends AggregateBase {
       else return
     }
 
+    // If an error was noticed before the mode could be set (like in the early lifecycle of the page), immediately set to FULL mode
+    if (this.mode === MODE.ERROR && this.errorNoticed) {
+      this.mode = MODE.FULL
+    }
+
```
Comment (cwli24) [inline src/features/session_replay/aggregate/index.js:169]: This already happens in L107-L111

File: src/features/session_replay/aggregate/index.js (hunk 12 above)
Comment (cwli24) [reply src/features/session_replay/aggregate/index.js:76]: Yeah these are both checked for before this module is imported.

File: src/features/session_replay/aggregate/index.js [hunk 15]
```diff
@@ -22,38 +22,41 @@ import { AggregateBase } from '../../utils/aggregate-base'
 // would be better to get this dynamically in some way
 export const RRWEB_VERSION = '2.0.0-alpha.8'
 
+export const avgCompression = 0.12
```
Comment (cwli24) [inline src/features/session_replay/aggregate/index.js:25]: Styling convention:
```suggestion
export const AVG_COMPRESSION = 0.12
```

File: src/features/session_replay/aggregate/index.js [hunk 16]
```diff
@@ -304,6 +326,7 @@ export class Aggregate extends AggregateBase {
   /** Abort the feature, once aborted it will not resume */
   abort () {
```
Comment (cwli24) [inline src/features/session_replay/aggregate/index.js:327]: [no action required]

There's actually quite a couple `abort` places in here, so I'll add an abortcontroller abort here in shared-channel on my next Trace PR and listen to that instead of just SESSION_RESET

File: tests/specs/session-replay/session-state.e2e.js (hunk 11 above)
Comment (metal-messiah) [reply tests/specs/session-replay/session-state.e2e.js:18]: ty, fixed

File: tests/specs/session-replay/session-state.e2e.js (hunk 10 above)
Comment (metal-messiah) [reply tests/specs/session-replay/session-state.e2e.js]: changed

File: tests/specs/session-replay/harvest.e2e.js [hunk 17]
```diff
@@ -1,8 +1,25 @@
...
+          cap: 1,
+          spa: 1,
+          loaded: 1,
+          sr: 1
+        })
+      })
+    })
+
     afterEach(async () => {
+      await browser.testHandle.clearScheduledReplies('bamServer')
```
Comment (patrickhousley) [inline tests/specs/session-replay/harvest.e2e.js:22]: The test handle is always destroyed after each test. What was the issue being seen that prompted the creating of this method `clearScheduledReplies`?

File: tests/specs/session-replay/helpers.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (metal-messiah) [reply tests/specs/session-replay/helpers.js:20]: Will revisit this another time

File: tests/specs/session-replay/ingest.e2e.js (hunk 3 above)
Comment (metal-messiah) [reply tests/specs/session-replay/ingest.e2e.js:50]: added two tests, one for 64kb dom, and one for 1mb dom

File: src/features/utils/instrument-base.js (hunk 7 above)
Comment (metal-messiah) [reply src/features/utils/instrument-base.js:113]: Consulted with larger group, this will be fine for LP.  May need to circle back for GA.  Added a dependency check in `featureDependencies.js` as suggested.

File: src/features/session_replay/aggregate/index.js (hunk 14 above)
Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:162]: This extra `errorNoticed` check is needed for the case where:

An error was thrown and noticed before `mode` could be set to "ERROR" (ie before SR loads, before the sampling)

File: src/features/session_replay/aggregate/index.js (hunk 16 above)
Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:320]: good call

File: tests/specs/session-replay/harvest.e2e.js (hunk 17 above)
Comment (metal-messiah) [reply tests/specs/session-replay/harvest.e2e.js:22]: Wasn't sure if needed tbh (for a test case that might not actually call the RUM call, was not sure what would happen to the scheduled reply)

The actual issue that first prompted was that I wanted to schedule a reply for every test (since nearly all SR tests need the SR: 1 value, which is 0 by default), except one, so for that I just put into a beforeEach, then clear the scheduled response for the 1 test 

https://github.com/newrelic/newrelic-browser-agent/pull/571/files/b116fe95a8b30ca0cdcf42363daee1b8b677af66..766ac0a02dce62607d52435132cef644a7f36fc4#diff-68f6284501cbb642ecebf419e8a964f8dbff379ebc68af9eeb639b8dc4fc4ef6R29-R41

We can do some more clean up if needed

File: src/features/session_replay/aggregate/index.js (hunk 12 above)
Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:71]: Not sure how well we can trust that or not, seeing as customers could import session replay agg directly themselves if they really wanted to.  Not that it matters much either way.  Will remove from agg.

Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:71]: I will say, having these in the feature enables some component tests that would not be as easy if those checks are solely abstracted to the middleman layer

**Comment by metal-messiah** [comment]
Wdio run after pr comment fixes

https://github.com/newrelic/newrelic-browser-agent/actions/runs/5260339285

File: tests/specs/session-replay/harvest.e2e.js [hunk 18]
```diff
@@ -0,0 +1,60 @@
+import { testRumRequest } from '../../../tools/testing-server/utils/expect-tests'
+import { config, getSR } from './helpers'
+
+export default (function () {
```
Comment (patrickhousley) [inline tests/specs/session-replay/harvest.e2e.js:4]: This might be a left-over from trying to group tests but not necessary.

File: tests/specs/session-replay/harvest.e2e.js (hunk 17 above)
Comment (patrickhousley) [reply tests/specs/session-replay/harvest.e2e.js:22]: Ah ok. Not against it. Just was curious if there was an issue with handles getting stuck in memory or something. You could use a function at the bottom of the test file that enables the reply and only call it in the tests that needed it. No difference to me though.

File: tools/wdio/plugins/testing-server/test-handle-connector.mjs [hunk 19]
```diff
@@ -50,6 +50,37 @@ export class TestHandleConnector {
...
+      }),
+      headers: { 'content-type': 'application/json' }
+    })
+  }
+
+  /**
+   * Schedules a reply to a server request
+   * @param {'assetServer'|'bamServer'} serverId Id of the server the request will be received on
+   * @param {ScheduledReply} scheduledReply The reply options to apply to the server request
+   */
```
Comment (patrickhousley) [inline tools/wdio/plugins/testing-server/test-handle-connector.mjs:73]:
```suggestion
  /**
   * Clears all scheduled replies for the given server.
   * @param {'assetServer'|'bamServer'} serverId Id of the server the request will be received on
   */
```

File: tools/testing-server/constants.js [hunk 20]
```diff
@@ -35,7 +35,5 @@ module.exports.rumFlags = {
   cap: 1,
   spa: 1,
   loaded: 1,
-  sr: 0, // this should be off, for now, if privacy.cookie_enabled is on (default) or Traces tests will fail
-  srf: 1,
-  sre: 1
+  sr: 1 // this should be off, for now, if privacy.cookie_enabled is on (default) or Traces tests will fail
```
Comment (patrickhousley) [inline tools/testing-server/constants.js:38]: Does this need to be changed back to `0`?

Comment (metal-messiah) [reply tools/testing-server/constants.js:38]: Yes thanks for catching that

File: src/features/session_replay/aggregate/index.js (hunk 12 above)
Comment (metal-messiah) [reply src/features/session_replay/aggregate/index.js:71]: can revisit later if seen as a problem, i don't see this as a blocker for now.

**Comment by metal-messiah** [comment]
Full JIL suite  --> https://github.com/newrelic/newrelic-browser-agent/actions/runs/5268215096/jobs/9526110766

File: tests/assets/rrweb-instrumented.html [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+      position: absolute; right: 50px; top: 200px;
+    }
+  </style>
+  {init}
+  {config}
+  {loader}
+  {script-injection}
+</head>
+<body>
+  this is a generic page that is instrumented by the JS agent
```
Comment (patrickhousley) [inline tests/assets/rrweb-instrumented.html:23]: Might change this to indicate this page has elements specific to rrweb implementation.

File: tests/assets/rrweb-record.html [hunk 22]
```diff
@@ -14,23 +14,12 @@
...
   <script>
-    NREUM.init.session_replay = {
-      enabled: true,
-      harvestTimeSeconds: 5,
-      sampleRate: 1,
-      errorSampleRate: 1
-    }
+    localStorage.clear()
+    // NREUM.init.session = {expiresMs: 5000}
+    NREUM.init.session_replay = {enabled: true, harvestTimeSeconds: 5, errorSampleRate: 0, sampleRate: 1}
```
Comment (patrickhousley) [inline tests/assets/rrweb-record.html:22]: Change the body text below to indicate this page is purposefully instrumented for rrweb local testing.

File: tests/specs/session-replay/harvest.e2e.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
+import { testRumRequest } from '../../../tools/testing-server/utils/expect-tests'
+import { config, getSR } from './helpers'
+
+describe('Session Replay Harvest Behavior', () => {
+  beforeEach(async () => {
+    await browser.testHandle.scheduleReply('bamServer', {
```
Comment (patrickhousley) [inline tests/specs/session-replay/harvest.e2e.js:6]: Maybe use `browser.addCommand('enableSessionReplay'...)`

**Review by patrickhousley — APPROVED** [review]
I think there might still be some rare edge case flaky-ness with some of the expect/promise ordering in some of the tests but if they are passing consistently, I think we can look at it again later if it becomes an issue.

**Review by bjfield — APPROVED** [review]
